consensus/src/simplex/signing_scheme/bls12381_multisig.rs

```diff
@@ -0,0 +1,1113 @@
+                "consensus::simplex::signing_scheme::bls12381_multisig::Certificate",
+                "Signers are not strictly increasing",
+            ));
+        }
+
+        if signers
+            .iter()
+            .any(|signer| (*signer as usize) >= *participants)
+        {
+            return Err(Error::Invalid(
+                "consensus::simplex::signing_scheme::bls12381_multisig::Certificate",
+                "Signer index exceeds participant set",
+            ));
+        }
+
+        let signature = V::Signature::read(reader)?;
+
+        Ok(Self { signers, signature })
+    }
+}
+
+impl<V: Variant + Send + Sync> signing_scheme::Scheme for Scheme<V> {
+    type Signature = V::Signature;
+    type Certificate = Certificate<V>;
+    type Seed = ();
+
+    fn sign_vote<D: Digest>(
+        &self,
+        namespace: &[u8],
+        context: VoteContext<'_, D>,
+    ) -> Option<Vote<Self>> {
+        let (index, private_key) = self.signer.as_ref()?;
+
+        let (namespace, message) = vote_namespace_and_message(namespace, context);
+        let signature = sign_message::<V>(private_key, Some(namespace.as_ref()), message.as_ref());
+
+        Some(Vote {
+            signer: *index,
+            signature,
+        })
+    }
+
+    fn verify_vote<D: Digest>(
+        &self,
+        namespace: &[u8],
+        context: VoteContext<'_, D>,
+        vote: &Vote<Self>,
+    ) -> bool {
+        let Some(public_key) = self.participant(vote.signer) else {
+            return false;
+        };
+
+        let (namespace, message) = vote_namespace_and_message(namespace, context);
+        verify_message::<V>(
+            public_key,
+            Some(namespace.as_ref()),
+            message.as_ref(),
+            &vote.signature,
+        )
+        .is_ok()
+    }
+
+    fn verify_votes<R, D, I>(
+        &self,
+        rng: &mut R,
+        namespace: &[u8],
+        context: VoteContext<'_, D>,
+        votes: I,
+    ) -> VoteVerification<Self>
+    where
+        R: Rng + CryptoRng,
+        D: Digest,
+        I: IntoIterator<Item = Vote<Self>>,
+    {
+        let (namespace, message) = vote_namespace_and_message(namespace, context);
+        let message_hash =
+            hash_message_namespace::<V>(V::MESSAGE, namespace.as_ref(), message.as_ref());
+
+        let mut invalid = BTreeSet::new();
+        let mut candidates = Vec::new();
+
+        let mut publics = Vec::new();
+        let mut hms = Vec::new();
+        let mut signatures = Vec::new();
+
+        for vote in votes.into_iter() {
+            let Some(public_key) = self.participant(vote.signer) else {
+                invalid.insert(vote.signer);
+                continue;
+            };
+
+            publics.push(*public_key);
+            hms.push(message_hash);
+            signatures.push(vote.signature);
+
+            candidates.push(vote);
+        }
+
+        if !candidates.is_empty() && V::batch_verify(rng, &publics, &hms, &signatures).is_err() {
```

Comment:
we should use `aggregate_verify_multiple_public_keys` (much much more efficient than `batch_verify`)